helix-view/src/editor.rs

```diff
@@ -579,20 +579,41 @@ impl Editor {
             tokio::spawn(language_server.text_document_did_close(doc.identifier()));
         }
 
-        let views_to_close = self
+        enum Action {
+            Close(ViewId),
+            ReplaceDoc(ViewId, DocumentId),
+        }
+
+        let actions: Vec<Action> = self
             .tree
-            .views()
+            .views_mut()
             .filter_map(|(view, _focus)| {
                 if view.doc == doc_id {
-                    Some(view.id)
+                    // something was previously open in the view, switch to previous doc
+                    if let Some(prev_doc) = view.docs_access_history.pop() {
+                        Some(Action::ReplaceDoc(view.id, prev_doc))
+                    } else {
+                        // only the document that is being closed was in the view, close it
+                        Some(Action::Close(view.id))
+                    }
                 } else {
                     None
                 }
             })
-            .collect::<Vec<_>>();
+            .collect();
 
-        for view_id in views_to_close {
-            self.close(view_id);
+        for action in actions {
+            match action {
+                Action::Close(view_id) => {
+                    self.close(view_id);
+                }
+                Action::ReplaceDoc(view_id, doc_id) => {
+                    self.replace_document_in_view(view_id, doc_id);
+
+                    let view = self.tree.get_mut(view_id);
+                    view.jumps.remove(&doc_id);
```

Comment:
> > example document in the history being closed
> 
> That's a bug, such documents need to be removed here:
> 
> https://github.com/helix-editor/helix/blob/2a1fa0f8f9ae1e5df4822c88c9b7ed2faa2cbefc/helix-view/src/view.rs#L62-L64
> 
> This is how we clean up the jumplist when a document is closed.

Added here, does that make sense?

Reply:
The cleanup needs to be ran with the document id that was just closed and you want to run this on _all_ views, removing the id from both the jumplist and the access history.

Reply:
Done.